app/screens/custom_status/custom_status_modal.tsx

```diff
@@ -0,0 +1,391 @@
+
+type DefaultUserCustomStatus = {
+    emoji: string;
+    message: string;
+    messageDefault: string;
+};
+
+const defaultCustomStatusSuggestions: DefaultUserCustomStatus[] = [
+    {emoji: 'calendar', message: t('custom_status.suggestions.in_a_meeting'), messageDefault: 'In a meeting'},
+    {emoji: 'hamburger', message: t('custom_status.suggestions.out_for_lunch'), messageDefault: 'Out for lunch'},
+    {emoji: 'sneezing_face', message: t('custom_status.suggestions.out_sick'), messageDefault: 'Out sick'},
+    {emoji: 'house', message: t('custom_status.suggestions.working_from_home'), messageDefault: 'Working from home'},
+    {emoji: 'palm_tree', message: t('custom_status.suggestions.on_a_vacation'), messageDefault: 'On a vacation'},
+];
+
+interface Props extends NavigationComponentProps {
+    intl: typeof intlShape;
+    theme: Theme;
+    customStatus: UserCustomStatus;
+    recentCustomStatuses: UserCustomStatus[];
+    isLandscape: boolean;
+    actions: {
+        setCustomStatus: (customStatus: UserCustomStatus) => void;
+        unsetCustomStatus: () => void;
+        removeRecentCustomStatus: (customStatus: UserCustomStatus) => void;
+    };
+}
+
+type State = {
+    emoji: string;
+    text: string;
+}
+
+class CustomStatusModal extends NavigationComponent<Props, State> {
+    rightButton: OptionsTopBarButton = {
+        id: 'update-custom-status',
+        testID: 'custom_status.done.button',
+        enabled: true,
+        showAsAction: 'always',
+    };
+
+    static options(): Options {
+        return {
+            topBar: {
+                title: {
+                    alignment: 'center',
+                },
+            },
+        };
+    }
+
+    constructor(props: Props) {
+        super(props);
+
+        this.rightButton.text = props.intl.formatMessage({id: 'mobile.custom_status.modal_confirm', defaultMessage: 'Done'});
+        this.rightButton.color = props.theme.sidebarHeaderTextColor;
+
+        const options: Options = {
+            topBar: {
+                rightButtons: [this.rightButton],
+            },
+        };
+
+        mergeNavigationOptions(props.componentId, options);
+
+        this.state = {
+            emoji: props.customStatus.emoji,
+            text: props.customStatus.text || '',
+        };
+    }
+
+    componentDidMount() {
+        Navigation.events().bindComponent(this);
+    }
+
+    navigationButtonPressed() {
+        this.handleSetStatus();
+    }
+
+    handleSetStatus = () => {
+        const {emoji, text} = this.state;
+        const isStatusSet = emoji || text;
+        const {customStatus} = this.props;
+        if (isStatusSet) {
+            const isStatusSame = customStatus.emoji === emoji && customStatus.text === text;
+            if (!isStatusSame) {
+                const status = {
+                    emoji: emoji || 'speech_balloon',
+                    text: text.trim(),
+                };
+                this.props.actions.setCustomStatus(status);
```

Comment:
here you could determine if this was successful or not, same with clearing.. then react appropriately in the places where you were tracking the request.

Reply:
We can do this while clearing the status but I'm not sure how we can do it while setting the status. If we determine this was unsuccessful here how can we react appropriately in the settings sidebar?

Reply:
JavaScript allows the use of Events, that way you can respond to them without having to run mapStateToProps for every mounted component as you track a request. Look for EventEmitter in the codebase for examples.